perllib/FixMyStreet/PhotoStorage/FileSystem.pm

```diff
@@ -0,0 +1,124 @@
+package FixMyStreet::PhotoStorage::FileSystem;
+
+use Moose;
+use parent 'FixMyStreet::PhotoStorage';
+
+use Path::Tiny 'path';
+
+
+has upload_dir => (
+    is => 'ro',
+    lazy => 1,
+    default => sub {
+        my $dir = FixMyStreet->config('PHOTO_STORAGE_OPTIONS')->{UPLOAD_DIR} ||
+                  FixMyStreet->config('UPLOAD_DIR');
+        return path($dir)->absolute(FixMyStreet->path_to());
+    },
+);
+
+=head2 init
+
+Creates UPLOAD_DIR and checks it's writeable.
+
+=cut
+
+sub init {
+    my $self = shift;
+    my $cache_dir = $self->upload_dir;
+    $cache_dir->mkpath;
+    unless ( -d $cache_dir && -w $cache_dir ) {
+        warn "\x1b[31mCan't find/write to photo cache directory '$cache_dir'\x1b[0m\n";
+        return;
+    }
+    return 1;
+}
+
+=head2 get_file
+
+Returns a Path::Tiny path to a file on disk identified by an ID and type.
+File may or may not exist. This handle is then used to read photo data or
+write to disk.
+
+=cut
+
+sub get_file {
+    my ($self, $fileid, $type) = @_;
+    my $cache_dir = $self->upload_dir;
+    return path( $cache_dir, "$fileid.$type" );
+}
+
+
+=head2 store_photo
+
+Stores a blob of binary data representing a photo on disk.
+Returns a key which is used in the future to get the contents of the file.
+
+=cut
+
+sub store_photo {
+    my ($self, $photo_blob) = @_;
+
+    my $type = $self->detect_type($photo_blob) || 'jpeg';
+    my $fileid = $self->get_fileid($photo_blob);
+    my $file = $self->get_file($fileid, $type);
+    $file->spew_raw($photo_blob);
```

Comment:
Guess you mean `copy_to`, from the original `$upload->copy_to( $file );`?
Not sure if there's much to be gained by reworking things to handle that, as the rotated photo contents has already been read in to memory above (`my $photo = $upload->slurp;`) and whilst copying the file on disk is perhaps marginally faster than writing from memory, is it going to be enough to be noticed? I like the simpler "here's a blob of photo data, please do something with it" API rather than "here's something which might be a blob or it might be something else, please figure it out and store it".

Reply:
Fair enough!